### ISSUE TITLE
Speicherzugriffsfehler (Speicherabzug geschrieben)

### ISSUE THREAD

**Issue by AndiLeni** [xcsf-dev/xcsf#185]
Hello,

I have some trouble using the library with real values inputs.
On my ubuntu machine I receive a "Speicherzugriffsfehler (Speicherabzug geschrieben)" with my RL agent.

The weird thing is, I am not able to reliably recreate this error. 
The code seems to work fine on my windows 10 laptop, but segfaults in random? runs on linux.
Sometimes it seems like calling `py3clean` helps and makes it run afterwards. Sometimes is doesn't.
Disabling bytecode caching via 
````
import sys
sys.dont_write_bytecode = True
````
did also not help.

Since I am running multiple epxeriments in parallel I thought maybe this might be an issue, but it also segfaults when running only one experiment at a time. 

Using the same agent but configured to use a binary input using 
```
condition = {
    "type": "ternary",
    "args": {
        "bits": 1,
        "p_dontcare": 0.5,
    },
}
````
works fine without an issue.

Is there some way to debug the cause of this behavior?

Thanks in advance and kind regards
Andreas

Here is the agents code as a reference:
```python
import json
import numpy as np

from xcsf import XCS


class Observation:
    def __init__(
        self,
        obs: list = None,
        reward: int = None,
        terminated: bool = None,
        truncated: bool = None,
        info: dict = None,
    ):
        self.state: list = obs
        self.info: dict = info
        self.reward: int = reward
        self.terminated: bool = terminated
        self.truncated: bool = truncated

    def __repr__(self):
        return f"Observation(state={self.state}, reward={self.reward}, terminated={self.terminated}, truncated={self.truncated}, info={self.info})"


class XcsrAgent:
    def __init__(self, gamma, epsilon, env, e0=0.01, beta=0.1, pop_size=2000):

        self.gamma = gamma
        self.epsilon = epsilon
        self.epsilon_min = 0.1
        self.e0 = e0
        self.beta = beta
        self.pop_size = pop_size

        self.old_state = None
        self.old_action = None

        # https://github.com/xcsf-dev/xcsf/wiki/Python-Library-Usage#hyperrectangles-and-hyperellipsoids
        # https://github.com/xcsf-dev/xcsf/wiki/Python-Library-Usage#initialising-actions
        # https://github.com/xcsf-dev/xcsf/wiki/Python-Library-Usage#initialising-predictions

        self.model = XCS(
            x_dim=6,
            y_dim=1,
            # n_actions=env.get_action_dim(),
            n_actions=239,
            gamma=0.9,
            e0=0.01,
            beta=0.1,
            pop_size=2000,
            condition={
                "type": "hyperrectangle_csr",  # center-spread
                "args": {
                    "min": 0,  # minimum value of a center/bound
                    "max": 1,  # maximum value of a center/bound
                    "spread_min": 0.1,  # minimum initial spread
                    "eta": 0,  # gradient descent rate for moving centers to mean inputs matched
                },
            },
            pop_init=True,
            random_state=21,
            omp_num_threads=1,
            prediction={
                "type": "constant",  # Constant predictions
            },
            alpha=0.1,
            action={
                "type": "integer",  # Integer actions for discrete action selection
            },
        )

    def get_agent_info(self):
        return {
            "Agent name": "XCS Agent",
            "Epsilon": self.epsilon,
            "Gamma": self.gamma,
            "e0": self.e0,
            "Beta": self.beta,
            "Population size": self.pop_size,
        }

    def get_action(self, obs):
        # normalize state
        obs.state = self._min_max_scale(obs.state)

        self.old_state = obs.state
        if np.random.uniform(0, 1) <= self.epsilon:
            self.old_action = self.model.decision(np.array(self.old_state), True)
        else:
            self.old_action = self.model.decision(np.array(self.old_state), False)

        return self.old_action

    def update(self, obs):
        self.model.update(obs.reward, obs.terminated)
        error = self.model.error(obs.reward, obs.terminated, 0)
        return error

    def save_model(self, path):
        self.model.save(path + ".bin")

    def load_model(self, path):
        self.model.load(path + ".bin")

    def _min_max_scale(self, data: list) -> list:
        """
        Scale the state to the range [0, 1].
        See: https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html and https://www.oreilly.com/library/view/feature-engineering-made/9781787287600/aa5580ee-6fb7-4ac2-a1fe-369d95b70168.xhtml
        Output: numpy arrays of type numpy.float64
        """
        min_val = min(data)
        max_val = max(data)

        if max_val == min_val:
            return [0.0] * len(data)

        diff = max_val - min_val
        x_scaled = [(x - min_val) / (diff) for x in data]

        return x_scaled

```

**Comment by AndiLeni** [comment]
Here is the stacktrace from gdb if it helps in any way:
```
= Running experiment: 1intersection_xcsr_single with seed: 42
Loop iteration: 1/200
Training ...
Episode: 1/20

Thread 1 "python" received signal SIGSEGV, Segmentation fault.
0x00007fffb3e7d097 in clset_match () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
(gdb) bt
#0  0x00007fffb3e7d097 in clset_match () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
#1  0x00007fffb3e93fb7 in xcs_rl_decision () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
#2  0x00007fffb3e62827 in ?? () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
#3  0x00007fffb3e6c400 in ?? () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
#4  0x00007fffb3e5ad1b in ?? () from /home/oclehrstuhl/Schreibtisch/sumo/.venv/lib/python3.12/site-packages/xcsf/xcsf.cpython-312-x86_64-linux-gnu.so
#5  0x000000000054a9f4 in cfunction_call (func=0x7fff4a7e4900, args=0x7fff4936d140, kwargs=0x0) at /usr/local/src/conda/python-3.12.4/Objects/methodobject.c:537
#6  0x000000000051b30b in _PyObject_MakeTpCall (tstate=0x9c11f8 <_PyRuntime+459704>, callable=0x7fff4a7e4900, args=<optimized out>, nargs=<optimized out>,
    keywords=0x0) at /usr/local/src/conda/python-3.12.4/Objects/call.c:240
#7  0x0000000000525e75 in _PyEval_EvalFrameDefault (tstate=<optimized out>, frame=0x7ffff7fb23c8, throwflag=<optimized out>) at Python/bytecodes.c:2714
#8  0x00000000005e4dfe in PyEval_EvalCode (co=co@entry=0xa88430, globals=globals@entry=0x7ffff75f91c0, locals=locals@entry=0x7ffff75f91c0)
    at /usr/local/src/conda/python-3.12.4/Python/ceval.c:578
#9  0x000000000060a5d7 in run_eval_code_obj (tstate=tstate@entry=0x9c11f8 <_PyRuntime+459704>, co=co@entry=0xa88430, globals=globals@entry=0x7ffff75f91c0,
    locals=locals@entry=0x7ffff75f91c0) at /usr/local/src/conda/python-3.12.4/Python/pythonrun.c:1722
#10 0x0000000000605ca7 in run_mod (mod=mod@entry=0xa8e210, filename=filename@entry=0x7ffff752fe70, globals=globals@entry=0x7ffff75f91c0,
    locals=locals@entry=0x7ffff75f91c0, flags=flags@entry=0x7fffffffdd20, arena=arena@entry=0x7ffff751bc90)
    at /usr/local/src/conda/python-3.12.4/Python/pythonrun.c:1743
#11 0x000000000061db12 in pyrun_file (fp=fp@entry=0x9fc730, filename=filename@entry=0x7ffff752fe70, start=start@entry=257, globals=globals@entry=0x7ffff75f91c0,
    locals=locals@entry=0x7ffff75f91c0, closeit=closeit@entry=1, flags=0x7fffffffdd20) at /usr/local/src/conda/python-3.12.4/Python/pythonrun.c:1643
#12 0x000000000061d3b0 in _PyRun_SimpleFileObject (fp=0x9fc730, filename=0x7ffff752fe70, closeit=1, flags=0x7fffffffdd20)
    at /usr/local/src/conda/python-3.12.4/Python/pythonrun.c:433
#13 0x000000000061d123 in _PyRun_AnyFileObject (fp=0x9fc730, filename=filename@entry=0x7ffff752fe70, closeit=closeit@entry=1, flags=flags@entry=0x7fffffffdd20)
    at /usr/local/src/conda/python-3.12.4/Python/pythonrun.c:78
#14 0x0000000000615b43 in pymain_run_file_obj (skip_source_first_line=0, filename=0x7ffff752fe70, program_name=0x7ffff752ff90)
    at /usr/local/src/conda/python-3.12.4/Modules/main.c:360
#15 pymain_run_file (config=0x963dd8 <_PyRuntime+77720>) at /usr/local/src/conda/python-3.12.4/Modules/main.c:379
#16 pymain_run_python (exitcode=0x7fffffffdcf4) at /usr/local/src/conda/python-3.12.4/Modules/main.c:629
#17 Py_RunMain () at /usr/local/src/conda/python-3.12.4/Modules/main.c:709
#18 0x00000000005cc259 in Py_BytesMain (argc=<optimized out>, argv=<optimized out>) at /usr/local/src/conda/python-3.12.4/Modules/main.c:763
#19 0x00007ffff7c2a1ca in __libc_start_call_main (main=main@entry=0x5cc190 <main>, argc=argc@entry=6, argv=argv@entry=0x7fffffffdf88)
    at ../sysdeps/nptl/libc_start_call_main.h:58
#20 0x00007ffff7c2a28b in __libc_start_main_impl (main=0x5cc190 <main>, argc=6, argv=0x7fffffffdf88, init=<optimized out>, fini=<optimized out>,
    rtld_fini=<optimized out>, stack_end=0x7fffffffdf78) at ../csu/libc-start.c:360
#21 0x00000000005cc089 in _start ()
```

**Comment by rpreen** [comment]
It looks like you are doing multi-step RL and want action sets like classic XCS.

The main way for this is method 1 on the wiki; see the maze example, which requires more explicit set initialisation by calling `init_trial()` to initialise the sets for a new episode and then looping the steps with `init_step()` and `end_step()`  - these let you work with `decision()` and `update()`

https://github.com/xcsf-dev/xcsf/blob/264c6310a378683be0ee86da026a5f97d1a1f9db/python/example_maze.py#L238-L256

A possible different way:

Method 2 is meant for single-step RL and is shown in the rmux example, which uses the function `fit(state, action, reward)` - this function will call all of the initialisation and ending functions automatically and create and update the action set. Obviously this doesn't handle the discounting of reward, so you would need to compute that in Python and then update the sets appropriately - how you do this depends on whether your learning has to occur online or can be done like with experience replay; the cartpole example shows something similar, but it's not using action sets. It would be a similar scheme to that but instead calling the `fit(state, action, reward)` to direct the reward to the appropriate action set, whereas in that cartpole example it's just calling `fit(state, reward)` to update the match set since it has classifiers with prediction vectors of length N_ACTIONS, representing the predicted payoff for every action like with deepQ nets. This may sound a bit too confusing without an example, so the easiest is just method 1.

**Comment by rpreen** [comment]
If you're already calling the `init_trial()` `init_step()` `end_step()` `end_trial()` functions and I just don't see that, then it's clearly crashing in the `clset_match()` function trying to create the match set, so i would start by verifying your inputs are scaled [0,1] and that the length of the vectors is 6 (the same as `x_dim` in the constructor). I would also be careful of the population size because with 239 actions that is a lot of different actions it will need to cover for and you will want some diversity in those sets, so you may need a much bigger population, especially with hyperrectangles. It's very hard to say without a complete minimal working/crashing example. I hope this helps.

**Comment by AndiLeni** [comment]
Yes, this did help! 
The init_ and end_ functions are handled by a runner script shich uses a helper to call these functions.
The helper however had a hardcoded classname to check for, which was not triggered by the xcsr agent.
I would not have checked there without your hint to take a look at these functions, so you definitely saved me a couple hours! 
Thank you very much!
